controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -1974,21 +1974,26 @@ private SegmentRange convert(String scope, String stream, Map.Entry<Long, Map.En
         }
     }
 
-    public CompletableFuture<Map<Long, Long>> notifyTxnCommit(final String scope, final String stream,
-                                                   final List<Long> segments, final UUID txnId, long requestId) {
+    public CompletableFuture<Map<Long, List<Long>>> mergeTxnSegmentsIntoStreamSegments(final String scope, final String stream,
+                                                                                       final List<Long> segments, final List<UUID> txnIds, long requestId) {
         Timer timer = new Timer();
         return Futures.allOfWithResults(segments.stream()
-                .collect(Collectors.toMap(x -> x, x -> notifyTxnCommit(scope, stream, x, txnId, requestId))))
-                .whenComplete((r, e) -> TransactionMetrics.getInstance().commitTransactionSegments(timer.getElapsed()));
+                        .collect(Collectors.toMap(segId -> segId, segId -> mergeTxnSegments(scope, stream, segId, txnIds, requestId))))
+                .whenComplete((r, e) -> {
+                    if (e != null) {
+                        Duration elapsed = timer.getElapsed();
+                        TransactionMetrics.getInstance().commitTransactionSegments(elapsed);
+                    }
+                });
     }
 
-    private CompletableFuture<Long> notifyTxnCommit(final String scope, final String stream,
-                                                    final long segmentNumber, final UUID txnId, long requestId) {
-        return TaskStepsRetryHelper.withRetries(() -> segmentHelper.commitTransaction(scope,
+    private CompletableFuture<List<Long>> mergeTxnSegments(final String scope, final String stream,
+                                                                             final long segmentNumber, final List<UUID> txnIds, long requestId) {
+        return TaskStepsRetryHelper.withRetries(() -> segmentHelper.mergeTxnSegments(scope,
```

Comment:
`segmentHelper.mergeTxnSegments(..)` will throw a `WireCommandFailedException(type, WireCommandFailedException.Reason.SegmentDoesNotExist)` in case of a no-such segment response from the SSS. This exception implements the RetryableException and the code would retry unnecessarily here, right ?

Reply:
`WireCommandFailedException.Reason.SegmentDoesNotExist` is a retryable exception because we'd want to retry when we get NoSuchSegment. This pattern is also followed by other commands like updateTableEntries(), removeTableKeys can receive NoSuchSegment and do retries on this case.
Also for this particular wire command NoSuchSegment is thrown only when the TargetSegment does not exist and not if the source(txn) segment does not exist.

Reply:
NoSuchSegment response from SSS implies the TxnSegment is already merged and there is no need to retry, right?

Reply:
My conjecture is it could happen because of other reasons as well which is why we're also retry on NoSuchSegment for other commands.

Reply:
I don't understand the above reasoning. If the segment doesn't exist are you expecting it to be created in the future, and the retries will work on the future segment? Or are you assuming the server is lying? I have the same question for updateTableEntries and removeTableKeys

Reply:
@RaulGracia My understanding is SSS returns a NoSuchSegment when the segment actually does **not** exist. 
Or is there a possibility that SSS could return a NoSuchSegment due to the Container not having that segment or such other reason, when the Segment actually does exist but on another SSS?

Reply:
@pbelgundi a `NoSuchSegment` is sent to the client when the Segment does not exist in the Container. The existence of a Segment on a container is checked via the metadata, which is essentially a `TableSegment` that every Segment Container has with all the info of Segments (on `TableEntry` per Segment). If there is no entry for the requested Segment in the metadata, then a `StreamSegmentNotExistsException` is thrown in the Segment Store, which is translated in a `NoSuchSegment` reply to the client. This is the piece of code where the `getSegmentInfo()` can throw that exception if the metadata of a Segment does not exist: https://github.com/pravega/pravega/blob/c9e3a165c1708872ecbac75fd2f127f57dcd0ade/segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/TableMetadataStore.java#L161

This the most common situation would be if a request asks by a Segment that has been already deleted or not yet created in that particular Segment Container. But given that Segments are routed by name to a Segment Container, I don't know how it can be possible that a client can get a `NoSuchSegment` if the Segment actually exists but the request goes to the wrong Segment Container.

Reply:
> NoSuchSegment response from SSS implies the TxnSegment is already merged and there is no need to retry, right?

For this particular wirecommand - MergeSegmentsBatch, NoSuchSegment is returned by PravegaRequestProcessor only when the **TargetSegment** does not exist and never for the source (txn) segment.

Reply:
@RaulGracia based on your answer above I think _NoSuchSegment_ implies the Segment does **not** exist anywhere on Segment Store and so there is no point in retrying. Thanks for confirming.
I will change the behavior of this method accordingly. As for the older methods, I have created a separate issue to handle those: https://github.com/pravega/pravega/issues/6488

Reply:
Fixed. Changed the code to not retry on NoSuchSegment